### ISSUE TITLE
实验

### ISSUE THREAD

**Issue by zxs1995** [ShomyLiu/pytorch-relation-extraction#1]
您好，请问您的这份实现有test部分吗，不是predict的

**Comment by zxs1995** [comment]
您好，你的实验结果链接页面看不了，能否在重新发一下链接呢，谢谢

**Comment by ShomyLiu** [comment]
@zxs1995  我这里可以访问: http://shomy.top/2018/07/05/pytorch-relation-extraction/   如果访问异常，可以访问备用网址: https://blog.htliu.cn/2018/07/05/pytorch-relation-extraction/

**Comment by zxs1995** [comment]
麻烦您了，谢谢~

**Comment by zxs1995** [comment]
不好意思，您的FilterNYT的train和test数据具体格式是什么，能否说明一下呢，那个没太看懂，麻烦了~

**Comment by zxs1995** [comment]
不好意思，由于自己刚刚接触远程监督，所以想通过您复现的代码学习一下，但是有一部分没有看懂，想请教一下您。
（1）您的数据集
![image](https://user-images.githubusercontent.com/26573634/42980317-6a3c5eda-8c09-11e8-89e7-c3781834559e.png)
label的部分比如0 -1 -1 -1 2， 0 -1 -1 -1 我并未搞清楚这些是什么意思，还有就是我看了程序的代码，您是只选用了每行label的第一个是吗，按上面的例子就是0,这是实体对对应的label？
（2）我一直并未搞明白，论文中的评测结果是通过预测test集的label然后再和test集中正确的label进行比较计算初P和R值吗
（3）数据预处理的部分，您对数据进行padding，
![image](https://user-images.githubusercontent.com/26573634/42980852-c533df0a-8c0b-11e8-93c6-e7878636f5dc.png)
您的padding是按照最大的句长左右都padding了一个单位，为什么呢
麻烦了！

**Comment by ShomyLiu** [comment]
@zxs1995  这份数据预处理的代码，是参考PCNN的Theano代码来写的https://github.com/smilelhh/ds_pcnns，包括数据格式
- 数据描述
第一行: 两个实体的ID:  ent1_id  ent2_id
第二行前4个 是 label， -1表示没有。 第五个数是该bag内的句子数量N。因为有的bag可能有多个label，但是最多有4个，所以用了4个数来表示，比如 1 2 -1 -1 表示该bag存在1和2两种标签。训练的时候，的确是仅仅取了第一个作为训练句子。
后面N行跟该bag的句子
- PR值的计算：在预测的时候就是对test集的label， 但是计算PR的时候，需要去掉无意义的NA相互对比类型，即：预测为NA标签也是NA的部分。

- 进行Padding的原因是后续因为CNN卷积保持宽度不变，是常用技巧

**Comment by zxs1995** [comment]
非常感谢您的解答，但是有一些点我还不是很清楚
(1)您代码中padding到max_len这块我可以理解，但是好像随后又padding了2个单位，这块有点疑惑，不知道为什么又padding两个单位
(2)test集，有些实体对，我看label有两个，那这怎么判断model预测出来的label是该句子中的实体对正确的relation呢，还是test集的每个句子中的实体对的relation其实也是不确定的，并不是准确的relation，这块也是很疑惑
(3)multi-instance learning是通过找到与此bag选取的label关联最大的句子，然后通过
![image](https://user-images.githubusercontent.com/26573634/42994578-0dd54af2-8c41-11e8-863b-ac65b06ea4ac.png)g公式来最大化J，不断的更新网络中的参数吗，那也就是默认该句子对应的relation是正确的？
还有，我看您代码里用到的是 nn.CrossEntropyLoss()，但是这个函数不是求最小值吗，是改写了公式吗
问题有点多，劳烦您再解答一下~谢谢

**Comment by ShomyLiu** [comment]
@zxs1995  
- 后续的padding是因为PCNN需要Piece wise Pooling三段式, 考虑到一些特殊边界情况，比如entity1 在句子开头的情况。
- test里面有多个的，只要可以预测对一个即可。 当然现在也有一些工作做多标签分类。
- 对，默认是对的
- 一般的优化函数，比如SGD,Adadelta 都是求最小化的。公式的是优化最大似然，加个负号就是最小了， 在高斯分布假设下，是可以推导出与交叉熵是等价的

**Comment by res-03wade** [comment]
博主您好，请问您目前对哪些关系抽取的论文做了实验复现呢？github上对attention+blstm的实验复现结果只有77%，而且感觉好多复现实验都达不到原文所述的效果，但是又找不到问题在哪。像您平常做论文的话，基线系统对比实验是怎么做的呢？

**Comment by charlesfufu** [comment]
> @zxs1995 这份数据预处理的代码，是参考PCNN的Theano代码来写的[https://github.com/smilelhh/ds_pcnns，包括数据格式](https://github.com/smilelhh/ds_pcnns%EF%BC%8C%E5%8C%85%E6%8B%AC%E6%95%B0%E6%8D%AE%E6%A0%BC%E5%BC%8F)
> 
> * 数据描述
>   第一行: 两个实体的ID:  ent1_id  ent2_id
>   第二行前4个 是 label， -1表示没有。 第五个数是该bag内的句子数量N。因为有的bag可能有多个label，但是最多有4个，所以用了4个数来表示，比如 1 2 -1 -1 表示该bag存在1和2两种标签。（已经更新到README）
>   训练的时候，的确是仅仅取了第一个作为训练句子。
>   后面N行跟该bag的句子
> * PR值的计算：在预测的时候就是对test集的label， 但是计算PR的时候，需要去掉无意义的NA相互对比类型，即：预测为NA标签也是NA的部分。
> * 进行Padding的原因是后续因为CNN卷积保持宽度不变，是常用技巧
数据集合难道不需要原文章的数据了吗？不需要word embedding了吗？

**Comment by charlesfufu** [comment]
> @res-03wade 关于全监督类型的关系抽取, 我实现过PCNN(Zeng 2014)那一篇，大概到了82左右吧。 我这份代码实现的关于远程监督的关系抽取，不太一样。 baseline的结果，直接用已发表的论文中的结果。



> @zxs1995 这份数据预处理的代码，是参考PCNN的Theano代码来写的[https://github.com/smilelhh/ds_pcnns，包括数据格式](https://github.com/smilelhh/ds_pcnns%EF%BC%8C%E5%8C%85%E6%8B%AC%E6%95%B0%E6%8D%AE%E6%A0%BC%E5%BC%8F)
> 
> * 数据描述
>   第一行: 两个实体的ID:  ent1_id  ent2_id
>   第二行前4个 是 label， -1表示没有。 第五个数是该bag内的句子数量N。因为有的bag可能有多个label，但是最多有4个，所以用了4个数来表示，比如 1 2 -1 -1 表示该bag存在1和2两种标签。（已经更新到README）
>   训练的时候，的确是仅仅取了第一个作为训练句子。
>   后面N行跟该bag的句子
> * PR值的计算：在预测的时候就是对test集的label， 但是计算PR的时候，需要去掉无意义的NA相互对比类型，即：预测为NA标签也是NA的部分。
> * 进行Padding的原因是后续因为CNN卷积保持宽度不变，是常用技巧



数据集合难道不需要原文章的数据了吗？不需要word embedding了吗？

**Comment by ShomyLiu** [comment]
@charlesfufu  不太懂您的意思, 上面的链接的Theano代码 就是原文章PCNN+ONE（zeng 2015）公开的源代码以及数据集。

**Comment by charlesfufu** [comment]
> @charlesfufu 不太懂您的意思, 上面的链接的Theano代码 就是原文章PCNN+ONE（zeng 2015）公开的源代码以及数据集。

我的意思是不太理解数据集合，为什么没有包含实体的原文章信息

**Comment by ShomyLiu** [comment]
@zxs1995  有的。 数据集中做一个映射，word->id. 可以根据这个表找到原文章信息。

**Comment by qwerfdsaplking** [comment]
> 您好，请问您的这份实现有test部分吗，不是predict的

您好，我建了一个文本实体关系抽取的QQ群，有兴趣可以加入交流，711741227
